src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs

```diff
@@ -381,8 +395,24 @@ private static unsafe bool TryParseNumber(ref char* str, char* strEnd, NumberSty
             }
 
             bool negExp = false;
-            number.DigitsCount = digEnd;
+            if (digEnd == maxDigCount && numberOfTrailingZeros > 0 && number.HasNonZeroTail)
+            {
+                // We have an additional non-zero digit in the input past maxDigitCount. We already handle this properly.
+                number.DigitsCount = digEnd;
+            }
+            else
+            {
+                // We have only 0s in the fractional part. We can strip it all out
+                number.DigitsCount = digEnd - numberOfTrailingZeros;
+            }
             number.Digits[digEnd] = (byte)('\0');
```

Comment:
Shouldn't this be setting `number.Digits[number.DigitsCount]` so that the buffer is `5.123\0` and not `5.1230000\0`?

Reply:
It is now. (Initially I wanted to loop back over the buffer and zero out the trailing zero indices. That's all changed now)